scripts/pkg/build_templates/opensearch-dashboards/deb/debian/postinst

```diff
@@ -29,23 +29,32 @@ if command -v systemd-tmpfiles > /dev/null; then
     systemd-tmpfiles --create opensearch-dashboards.conf
 fi
 
-# Messages
-echo "### NOT starting on installation, please execute the following statements to configure opensearch-dashboards service to start automatically using systemd"
-echo " sudo systemctl daemon-reload"
-echo " sudo systemctl enable opensearch-dashboards.service"
-echo "### You can start opensearch-dashboards service by executing"
-echo " sudo systemctl start opensearch-dashboards.service"
+# Restart the service after an upgrade
+if [ "$1" == "configure" ] && [ "$2" != "" ]; then
+    if command -v systemctl >/dev/null && systemctl is-enabled opensearch-dashboards.service >/dev/null; then
```

Comment:
If the unit is enabled but not running, we don't want the upgrade to start it.
If the unit is disabled but the service is running, we want the upgrade to restart it.

```suggestion
    if command -v systemctl >/dev/null && systemctl is-active opensearch-dashboards.service >/dev/null; then
```

Same for the other services.

Reply:
Ah, the `prerm` script stop the service, so this would never work.  What a mess :unamused: …  My vote would be to NOT stop the services at the `prerm` stage, and `systemctl try-restart opensearch-dashboards.service` in the `postinst` stage.

Reply:
Hi,

@peterzhuamazon what are your thoughts on this? Technically I could implement it 2 ways:

1. Don't stop the service in prerm and preinst, when the command line arguments to those scripts show the package is being upgraded, then check if the service is active (or maybe both active and enabled) in postinst and restart it.
2. Stop the service in prerm/preinst but create a file (e. g. /tmp/opensearch-upgrade-flag), when the command line arguments show the package is being upgraded, then in postinst check if the file exists and only start the service if the file exists (and maybe additionally if the service is enabled).

Keep in mind, that if this is implemented, testing will either require to create 2 new package versions with the updated maintainer scripts or to replace /var/lib/dpkg/info/opensearch.prerm with the updated version before an upgrade is attempted.

Thanks

Reply:
If we dont stop the service in prerm, then it will result in a case where service is already started but necessary files get deleted, which might even cause corrupt indices in data.

If option 2, what is the way to decide if that is a upgrade? Or just removal? The problem with this approach is, if someone issue a removal of version 1, then later they just want to install version 2 from scratch, this /tmp file will treat install from scratch as upgrade.

Reply:
Also, if we remove the stop service from prerm, it will become a breaking change for sure.

Reply:
> If we dont stop the service in prerm, then it will result in a case where service is already started but necessary files get deleted, which might even cause corrupt indices in data.
> 
> 
> 
> If option 2, what is the way to decide if that is a upgrade? Or just removal? The problem with this approach is, if someone issue a removal of version 1, then later they just want to install version 2 from scratch, this /tmp file will treat install from scratch as upgrade.

The way to differentiate between an upgrade or removal in prerm is to check the passed command line arguments. If a package is removed the first command line argument is "remove", if a package is upgraded it is "upgrade". The tmp-file itself would be removed in postinst immediately after starting the service (if the file existed).

Reply:
> If we dont stop the service in prerm, then it will result in a case where service is already started but necessary files get deleted, which might even cause corrupt indices in data.

Ah!  Not a good idea then.

The concept of a "flag file" (option 2) seems really fragile indeed.

Nothing is fine, everything is terrible :fearful:

Reply:
Ok, then under these circumstances I'm afraid I don't know about a way to implement the functionality for a restart if the service was active before upgrading.

Reply:
@smortex what would be your ideal solution to this problem? Can you elaborate so we can look for optimal solution?

Reply:
I don't think there is a solution to this problem: Debian manage service a certain way, has all the tooling to make this straightforward, but we want a different behavior and it is a pain.

The current code that check if the service is enabled and stop / start it if so does not really make sense but is probably the least-worse that can be done.